File: Recursion/Lecture 15/LeetCode/Letter Combinations of a Phone Number.py
"""
This question requries us to print all Valid Combinations of the given string digits which comprises of digits from 2 to 9.
The only thing we require extra in this question is to map all the letter combinations of the phone number to its corresponding digits.
The rest part of the question remains exactly same

# Hypothesis:
takes digits as input then appends all possible combination of the digits and the maped strings in the particular digit.

digit == string

choise diagraam:
1. 1 char can be choosen only once if it's from one number 
2. after choosing 1 char from a digit move i pointer and search for next char in digit and add all combination of the chars in the next dgit.
3. Continue it till i pointer is == len(digits) *********** BASE CASE ************* 

"""

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        
        def helper(i):
            
            # If i has reached till the end of digits, we have reached a Valid Combination
            # We should include it into ans[][] and return back
            
            if i == len(digits):
                
                # converting the comb [] to a string, then storing it to ans [] | As it'll be more dificult to push and pop() in a string, becuause strings are immutable in python
                
                new = ""
                for j in comb:
                    new += j
                ans.append(new)
                return
            
            # j will take care for the iterating over each chars in numCharMap.values() in a particular key "digit[i]"
            
            for j in range(len(numCharMap[digits[i]])):
            
                # picking the jth char from ith digit from the numCharMap
                
                comb.append(numCharMap[digits[i]][j])
                
                #  Ask recursion to do rest of the task
                
                helper(i+1)
                
                # Backtrack and undo the change we have already done
                
                comb.pop()
                            
        
        # As per the problem, for an empty string, we need to return an empty array
        if not digits:
            return []
        
        # To solve this question, we need to map all the letters of a particular digit with it's corresponding digit. This is done with the help of a dictionary.
        
        numCharMap = {  "2" : "abc", 
                        "3" : "def", 
                        "4" : "ghi",
                        "5" : "jkl", 
                        "6" : "mno",
                        "7" : "pqrs",
                        "8" : "tuv",
                        "9" : "wxyz"}
        
        ans = []
        comb = []
        
        helper(0)
        
        return ans
    
"""
Time Complexity: O(3^N)
Space Complexity: O(N)
"""

File: Recursion/Lecture 15/LeetCode/Letter Combinations of a Phone Number.java
/* This question requries us to print all Valid Combinations of the given string digits which comprises of digits from 2 to 9.
The only thing we require extra in this question is to map all the letter combinations of the phone number to its corresponding digits.
The rest part of the question remains exactly same
*/

class Solution
{
    void help(int i, String s, String temp, List<String> ans, HashMap<Character, String> m)
    {
        // If i has reached till the end of given string, we have reached a Valid Combination
        // We should include it into ans and return back

        if (i == s.length())
        {
            ans.add(temp);
            return;
        }

        // Since, we have mapped all the digits with their corresponding Letters
        // m.get( s.charAt(i) ) will give us the actual string corresponding to that digit using HashMap m
        
        char chr = s.charAt(i);
        String str = m.get(chr);

        // loop for traversing every element in str

        for (int j = 0; j < str.length(); j++)
        {
            
            char c = str.charAt(j);

            // Ask recursion to do rest of the task

            help(i + 1, s, temp + c, ans, m);

        }
    }
    
    public List<String> letterCombinations(String digits) 
    {
        
        // To solve this question, we need to map all the letters of a particular digit with it's corresponding digit. This is done with the help of HashMap <Character,String> m.

        HashMap<Character,String> m = new HashMap<>();
        m.put('2',"abc");
        m.put('3',"def");
        m.put('4',"ghi");
        m.put('5',"jkl");
        m.put('6',"mno");
        m.put('7',"pqrs");
        m.put('8',"tuv");
        m.put('9',"wxyz");

        List<String> ans = new ArrayList<>();

        if(digits.length() == 0)
            return ans;
        
        String temp = "";

        help(0 , digits , temp , ans , m) ;

        return ans;
    }
}

/*
Time Complexity: O(3^N)
Space Complexity: O(N)
*/
